src/CallManager.cpp

```diff
@@ -0,0 +1,320 @@
+}
+
+void
+CallManager::handleEvent(const RoomEvent<CallInvite> &callInviteEvent)
+{
+  nhlog::ui()->debug("CallManager::incoming CallInvite from {} with id {}", callInviteEvent.sender, callInviteEvent.content.call_id);
+
+  if (callInviteEvent.content.call_id.empty())
+    return;
+
+  std::vector<RoomMember> members(cache::getMembers(callInviteEvent.room_id));
+  if (onActiveCall() || members.size() != 2) {
+    emit newMessage(QString::fromStdString(callInviteEvent.room_id),
+        CallHangUp{callInviteEvent.content.call_id, 0, CallHangUp::Reason::InviteTimeOut});
+    return;
+  }
+
+  playRingtone("qrc:/media/media/ring.ogg", true);
+  roomid_ = QString::fromStdString(callInviteEvent.room_id);
+  callid_ = callInviteEvent.content.call_id;
+  remoteICECandidates_.clear();
+
+  const RoomMember &caller = members.front().user_id == utils::localUser() ? members.back() : members.front();
+  emit newCallParty(caller.user_id, caller.display_name);
+
+  auto dialog = new dialogs::AcceptCall(caller.user_id, caller.display_name, MainWindow::instance());
+  connect(dialog, &dialogs::AcceptCall::accept, this,
+      [this, callInviteEvent](){
+        MainWindow::instance()->hideOverlay();
+        answerInvite(callInviteEvent.content);});
+  connect(dialog, &dialogs::AcceptCall::reject, this,
+      [this](){
+        MainWindow::instance()->hideOverlay();
+        hangUp();});
+  MainWindow::instance()->showSolidOverlayModal(dialog);
+}
+
+void
+CallManager::answerInvite(const CallInvite &invite)
+{
+  stopRingtone();
+  std::string errorMessage;
+  if (!session_.init(&errorMessage)) {
+    emit ChatPage::instance()->showNotification(QString::fromStdString(errorMessage));
+    hangUp();
+    return;
+  }
+
+  setTurnServers();
+  session_.setStunServer(settings_->useStunServer() ? STUN_SERVER : "");
+
+  if (!session_.acceptOffer(invite.sdp)) {
+    emit ChatPage::instance()->showNotification("Problem setting up call");
+    hangUp();
+    return;
+  }
+  session_.acceptICECandidates(remoteICECandidates_);
+  remoteICECandidates_.clear();
+}
+
+void
+CallManager::handleEvent(const RoomEvent<CallCandidates> &callCandidatesEvent)
+{
+  nhlog::ui()->debug("CallManager::incoming CallCandidates from {} with id {}", callCandidatesEvent.sender, callCandidatesEvent.content.call_id);
+  if (callid_ == callCandidatesEvent.content.call_id) {
+    if (onActiveCall())
+      session_.acceptICECandidates(callCandidatesEvent.content.candidates);
+    else {
+      // CallInvite has been received and we're awaiting localUser to accept or reject the call
+      for (const auto &c : callCandidatesEvent.content.candidates)
+        remoteICECandidates_.push_back(c);
+    }
+  }
+}
+
+void
+CallManager::handleEvent(const RoomEvent<CallAnswer> &callAnswerEvent)
+{
+  nhlog::ui()->debug("CallManager::incoming CallAnswer from {} with id {}", callAnswerEvent.sender, callAnswerEvent.content.call_id);
+  if (onActiveCall() && callid_ == callAnswerEvent.content.call_id) {
```

Comment:
I think there is some issue here still (maybe somewhere else):

1) Call from phone
2) Accept on web
3) Nheko still rings

Can you reproduce and fix that?

Reply:
Agreed, now fixed.